### PR TITLE
GPFQ

### PR DESCRIPTION
**Pull request by Giuseppe5**

### CODE REVIEW

File: src/brevitas_examples/imagenet_classification/ptq/ptq_common.py [hunk 1]
```diff
@@ -256,7 +256,7 @@ def apply_gptq(calib_loader, model, act_order=False):
     dtype = next(model.parameters()).dtype
     device = next(model.parameters()).device
     with torch.no_grad():
-        with gptq_mode(model, act_order=act_order, use_quant_activations=False) as gptq:
+        with gptq_mode(model, act_order=act_order, use_quant_activations=True) as gptq:
```
Comment (Giuseppe5) [inline src/brevitas_examples/imagenet_classification/ptq/ptq_common.py:259]: Restore to False

File: src/brevitas/graph/gpxq.py [hunk 2]
```diff
@@ -111,25 +93,25 @@ def __enter__(self):
...
                         f'Behaviour might deviate from what expected.')
 
                 # Attach hooks for GPTQ
                 if self._is_module_supported(module):
-                    gptq = GPTQ(
+                    gpxq = self.class_implementation(
                         module,
                         name,
-                        num_blocks=self.num_blocks,
+                        # num_blocks=self.num_blocks,
```
Comment (Giuseppe5) [inline src/brevitas/graph/gpxq.py:109]: remove

File: src/brevitas/graph/gpxq.py [hunk 3]
```diff
@@ -171,34 +198,84 @@ def catch_stopfwd(self, *args, **kwargs):
...
 
 
-class GPTQ():
+class gpfq_mode(gpxq_mode):
     """
-    Adapted from https://github.com/IST-DASLab/gptq, released under the following LICENSE:
+    Apply GPTQ algorithm https://arxiv.org/abs/2210.17323.
 
-    Copyright 2023 IST-DASLab
+    Args:
```
Comment (Giuseppe5) [inline src/brevitas/graph/gpxq.py:205]: Remove

File: src/brevitas/graph/gpxq.py [hunk 4]
```diff
@@ -417,47 +577,151 @@ def single_layer_update(self, percdamp=.01):
...
+        self.layer = layer
+        self.name = name
+        self.act_order = act_order
+
+        weight = layer.weight.data
+        self.layer.weight_orig_data = deepcopy(weight)
+
+        # By default, use groups = 1
+        self.groups = 1
+        if isinstance(self.layer, qnn.QuantConv2d):
```
Comment (volcacius) [inline src/brevitas/graph/gpxq.py:596]: is this right?

**Review by volcacius — COMMENTED** [review]
We should potentially account for updating float layers (e.g. last unquantized) based on the activation quantization error at the input.

**Review by volcacius — COMMENTED** [review]
For act_order, we could consider weights x activation magnitude as ordering criteria

File: src/brevitas/graph/gpxq.py [hunk 5]
```diff
@@ -417,47 +577,151 @@ def single_layer_update(self, percdamp=.01):
...
+                weight = weight.transpose(1, 0)  # This performs a view
+            weight = weight.flatten(1)
+
+        if self.groups == 1:
+            self.quant_input = self.quant_input[0]
+            self.float_input = self.float_input[0]
+
+            U = torch.zeros(weight.shape[0], self.float_input.shape[0], device=dev, dtype=dtype)
+            for t in range(weight.shape[1]):
+                U += weight[:, t].unsqueeze(1) * self.float_input[:, t].unsqueeze(0)
```
Comment (volcacius) [inline src/brevitas/graph/gpxq.py:698]: We need to experiment with replacing float_input with quant_input

File: src/brevitas/graph/gpxq.py [hunk 6]
```diff
@@ -417,47 +577,151 @@ def single_layer_update(self, percdamp=.01):
...
+        dev = weight.device
+        dtype = weight.dtype
+        if isinstance(self.layer, SUPPORTED_CONV_OP):
+            if isinstance(self.layer, (qnn.QuantConvTranspose1d, qnn.QuantConvTranspose2d)):
+                weight = weight.transpose(1, 0)  # This performs a view
+            weight = weight.flatten(1)
+
+        if self.groups == 1:
+            self.quant_input = self.quant_input[0]
+            self.float_input = self.float_input[0]
```
Comment (volcacius) [inline src/brevitas/graph/gpxq.py:694]: We need to experiment with replacing float_input with quant_input

**Comment by Giuseppe5** [comment]
> For act_order, we could consider weights x activation magnitude as ordering criteria

We need to define an order for quantizing the input channels. When multiplying weight x activation, the input channel is the inner dimension of the matmul which means that we "lose" that information

File: src/brevitas/graph/gpxq.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+
+class GPxQ(ABC):
+
+    def __init__(self, layer, name, act_order, parallel_layers=1) -> None:
+        self.layer = layer
+        self.name = name
+        self.act_order = act_order
+
+        weight = layer.weight.data
+        self.layer.weight_orig = deepcopy(layer.weight)
```
Comment (volcacius) [inline src/brevitas/graph/gpxq.py:140]: Add a flag to control this creation. Do a buffer that is initialized with detach().clone() and put it under a if that checks that it doesn't exist yet.
